Rename Twitter/X workflow task to match updated view names

davidpuplava opened the pull request:
This PR fixes the following error `Exception: The shape type 'UpdateTwitterStatusTask_Fields_Thumbnail' is not found` and similar exceptions when editing a workflow while the `X Integration` feature is enabled.

![image](https://github.com/OrchardCMS/OrchardCore/assets/5144137/748f480b-0f8d-4929-a7db-6420eaccc756)

## Steps to Reproduce
1. Pull latest version of `main` branch or any branch that contains commit `9e03d4f85267762e5446b42ad16d281e06f61f41`
2. Create a tenant or site with the blog recipe
3. Enable the following features: Workflows, X Integration, Sign in with X
4. Create new workflow called `test` (or whatever name)
5. Click Save
6. Expect workflow editor to display, but observe runtime exception

## Analysis
Looks to be related to the [this PR](https://github.com/OrchardCMS/OrchardCore/pull/15528) to rebrand Twitter to X. The view files for the Workflow Activity Task [were renamed](https://github.com/OrchardCMS/OrchardCore/pull/15528/commits/9e03d4f85267762e5446b42ad16d281e06f61f41#diff-2530d8b6de8fa4fff10e8a004516bbbdfd051a87738066ec8c74a1fc7b395dc5L41) but the task itself was not renamed.

## Solution
After I renamed the Task to match the view names, the exception disappears and I can add the Update Status task a workflow.

##  Considerations
- In this PR, I only updated the Task name and left other type names that include the word Twitter. If desired, I can update this PR to rename those as well, but not sure if that will have any adverse consequences.
- Renaming the `UpdateTwitterStatusTask` to ` UpdateXStatusTask` breaks existing Workflows that were created with the original `UpdateTwitterStatusTask` name. The workaround is to manually edit the database JSON to update the workflow definition to the new task name. Ideally, this change would happen automatically but I'm not sure how. This is why I [opened a discussion](https://github.com/OrchardCMS/OrchardCore/discussions/15928) on whether a DataMigration can be used to achieve this. In all honestly though, this Twitter/X may not be used enough to try and solve this automatically, especially since editing the database entry was trivial.

Fixes https://github.com/OrchardCMS/OrchardCore/issues/16349

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Refactor**
	- Renamed `UpdateTwitterStatusTask` to `UpdateXStatusTask` across various components in the OrchardCore.Twitter module to reflect updated functionality.
- **New Features**
	- Enhanced the Twitter module's workflow activities to support new status update tasks.
- **Bug Fixes**
	- Fixed issues related to the editing and updating of Twitter status tasks in the workflow drivers.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

davidpuplava commented:
@dotnet-policy-service agree

hishamco commented:
> Renaming the UpdateTwitterStatusTask to  UpdateXStatusTask breaks existing Workflows that were created with the original UpdateTwitterStatusTask name. The workaround is to manually edit the database JSON to update the workflow definition to the new task name. Ideally, this change would happen automatically but I'm not sure how. This is why I https://github.com/OrchardCMS/OrchardCore/discussions/15928 on whether a DataMigration can be used to achieve this. In all honestly though, this Twitter/X may not be used enough to try and solve this automatically, especially since editing the database entry was trivial.

Thanks, @davidpuplava for opening this PR, IMHO we either revert the task rename or create a migration to avoid breaking existing workflows, please create a database migration and let me know if you need help on this

davidpuplava commented:
> IMHO we either revert the task rename or create a migration to avoid breaking existing workflows

That's the right thing to do. I'll try and add a database migration to do that and will let you know if I have any questions.

sebastienros commented:
Yeah, this rename is not showing all the things it can break. I would revert everything IMO. As long as X doesn't sue us for using the Twitter name ;) We could also clone the module to use the X language and deprecate the other one, while providing an opt-in migration path to the new module until a version when we definitely remove it. Users could even do it manually at least since both would exist.

davidpuplava commented:
> please create a database migration and let me know if you need help on this

@hishamco I added a Data Migration but it doesn't seem to run even though I added it to Startup. Looks like the `GetFeaturesThatNeedUpdateAsync` method in the `DataMigrationManager` isn't picking up that the OrchardCore.Twitter feature needs to be updated. But given @sebastienros's about all the things that could break, I think I'll just rename the views back to `UpdateTwitterStatusTask_...` to resolve the exception.

> Yeah, this rename is not showing all the things it can break. I would revert everything IMO

Thanks for pointing this out @sebastienros. I'll revise this PR to change the view names back.

hishamco commented:
Please update the PR to show why the migration is not working, which might be a bug that needs a fix, otherwise, we will revert the rename change

davidpuplava commented:
> Please update the PR to show why the migration is not working

False alarm. I have the migrations working now. I overlooked that migration was already applied while testing. I've pushed up a data migration that renames the task in any workflow type definitions.

I tested on my existing tenant that had workflows created prior to rebranding. I confirmed that instances do not need a rename because they refer back to the workflow type by id.

I tested a blocking workflow that was in progress and that too does not seem to be a problem as @Piedone pointed out while answering [my question about how to migrate](https://github.com/OrchardCMS/OrchardCore/discussions/15928#discussioncomment-9301803).

Do recipes and deployments need to be considered for migrating?

I see that there is an IRecipeMigrator concept but I think that is more for recipes that are packaged with a module.

hishamco commented on src/OrchardCore.Modules/OrchardCore.Twitter/Migrations/TwitterMigrations.cs:
```diff
@@ -0,0 +1,74 @@
+            await using var connection = dbConnectionAccessor.CreateConnection();
+            await connection.OpenAsync();
+            await using var transaction = await connection.BeginTransactionAsync(session.Store.Configuration.IsolationLevel);
+            var dialect = session.Store.Configuration.SqlDialect;
```
I think this is wrong, no need to rely on specific data storage

davidpuplava replied:
> I think this is wrong, no need to rely on specific data storage

What do you mean? I think those variables are all using interfaces from System.Data.Common to avoid relying on a specific database implementation. Is there different way to update the Workflow item types aside from updating the database directly?

hishamco replied:
Ignore my above comment, seems I conflict with other PR :)

Piedone commented on src/OrchardCore.Modules/OrchardCore.Twitter/Startup.cs:
Please also do this in the Manifest: https://github.com/OrchardCMS/OrchardCore/pull/16350/files#diff-a9202c4e00f955210bba190b9f988d36c12ddbe384a63887b39add512f506e2e

Piedone commented on src/OrchardCore.Modules/OrchardCore.Twitter/Workflows/Activities/UpdateXStatusTask.cs:
Please also change the UI strings "Update X Status" everywhere to "Update X (Twitter) Status"

Piedone commented on src/OrchardCore.Modules/OrchardCore.Twitter/Migrations/TwitterMigrations.cs:
```diff
@@ -0,0 +1,67 @@
+        _shellDescriptor = shellDescriptor;
+    }
+
+    public int Create()
```
After using https://github.com/OrchardCMS/OrchardCore/pull/16350 to be able to add an `UpdateTwitterStatusTask`, I've switched to this branch to test the migrations. However, I still get:

![image](https://github.com/OrchardCMS/OrchardCore/assets/1976647/02c5abec-ed59-4928-a9fe-e6f7a7345d39)

davidpuplava replied:
Thanks @Piedone for the App_Data snapshot. The root cause was the Migration was skipped because it was wrapped in a call to `_shellDescriptor.WasFeatureAlreadyInstalled("OrchardCore.Twitter")` which returns false. I expected that this method would return true when the feature was installed but something related to the feature's serial number matching the shellDescriptor's serial number causes it to return false. The [code is here](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Abstractions/Shell/Extensions/ShellDescriptorExtensions.cs#L19).

I removed the call to that `WasFeatureAlreadyInstalled` method and the migration succeeded.

davidpuplava replied:
Ok, I pushed the latest changes to fix this and tested against your App_Data snapshot, @Piedone. Looks to work as expected now.
![image](https://github.com/user-attachments/assets/18580a49-18ec-44b5-835b-93518491873a)

Piedone commented on src/OrchardCore.Modules/OrchardCore.Twitter/Migrations/TwitterMigrations.cs:
```diff
@@ -0,0 +1,67 @@
+    {
+        if (_shellDescriptor.WasFeatureAlreadyInstalled("OrchardCore.Twitter"))
+        {
+            ShellScope.AddDeferredTask(async scope =>
```
Why is a deferred task necessary? This can be done just from within `Create`, migrations are for these kind of tasks. Then you don't need to manage the transaction either (what I don't think you need to do anyway).

davidpuplava replied:
> Why is a deferred task necessary?

No reason other than it's what I could find from other PRs that looked to migrate data:
- https://github.com/OrchardCMS/OrchardCore/pull/11052/files#diff-74a3ba1fb26c512218a56465ffdfdc50b501eebab27ba6b14ae96161968e8214R152
- https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs#L178
- https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Users/Migrations.cs#L252

> This can be done just from within Create

Sounds good - I'll look to move the migration code there.

Piedone replied:
This can be necessary if you want to make sure that other migrations run before this. I don't think this is an issue here.

davidpuplava replied:
> This can be necessary if you want to make sure that other migrations run before this. I don't think this is an issue here.

Ok, I refactored the code to remove the deferred task, and just update the database similar to what is found in the [documentation](https://docs.orchardcore.net/en/latest/reference/core/Data/).

Piedone commented on src/OrchardCore.Modules/OrchardCore.Twitter/Migrations/TwitterMigrations.cs:
```diff
@@ -0,0 +1,67 @@
+
+namespace OrchardCore.Twitter.Migrations;
+
+public class TwitterMigrations : DataMigration
```
```suggestion
public class TwitterToXMigrations : DataMigration
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Twitter/Workflows/Activities/UpdateXStatusTask.cs:
I merged https://github.com/OrchardCMS/OrchardCore/pull/16350 to fix the immediate issue. This will PR will be necessary for a better long-term solution. Then, please refer the cshtml renames from that PR.

davidpuplava replied:
Implemented as best as I can tell to use "X (Twitter)" where possible.

davidpuplava commented:
> @davidpuplava for instructions on how to resolve the merge conflicts due to https://github.com/OrchardCMS/OrchardCore/issues/16572 please follow the step listed in https://github.com/OrchardCMS/OrchardCore/issues/16572#issuecomment-2291836365.

Thanks @MikeAlhayek, your comment instructions worked well. One additional thing I had to do was resolve a merge conflict after the `git merge formatting...` step. The current state of my PR was that I had a deleted file which I needed to `get rm ...` and then complete the merge.

After that, everything built correctly, and all tests passed.

Piedone commented:
When you're done addressing all feedback of a review, click ["Re-request review"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#re-requesting-a-review) in the top-right corner for each reviewer when you're ready for another round of review, so they know that you're done.

davidpuplava commented:
I've tested against the X/Twitter v1.1 API and confirm the `UpdateXTwitterStatusTask` works as expected. I've re-requested reviews as mentioned I should do earlier.

But...I'm not certain that changing the workflow task type name (`UpdateXTwitterStatusTask.cs`) makes sense anymore. 

Changing the UI strings to "X (Twitter)" is beneficial for the user. Changing type names benefits developers, but the single letter company name X is  ambiguous enough that "XTwitter" is more descriptive anyway.

So the question stands, is the risk of changing the type name worth the value gained in simply adding and 'X' in front of Twitter.
Probably not.

I originally opened this PR only to get past the unhandled exception that was ultimately fixed with #16350 by changing the type names back.

That said, it's worth considering if this PR should be closed without merging.

Piedone reviewed (changes requested):
Checking this out again, I agree. My head hurts due to all the wasted effort to adapt to such a pointless rebranding, but I'd keep every technical identifier (file names, types, HTML IDs...) as they were (https://github.com/OrchardCMS/OrchardCore/pull/16350 reverted the shape template renames already). We shouldn't have done anything else than update UI labels. So, please in this PR:

- Only keep the UI string renames (but do keep all of those), having "(Twitter)" everywhere makes it possible to search for things.
- Even revert the `AdminMenuSignin` class/ID "X" rename, including in `NavigationItemText-x.Id.cshtml`, but do add "(Twitter)" here too.
- Check the module for other cases of these and do the same.

If you @hishamco @sebastienros object to this please comment.

hishamco commented:
> My head hurts due to all the wasted effort to adapt to such a pointless rebranding

Zoltan I think we did similar things in both Azure AD & Facebook Meta

Piedone commented:
Yes, and that was a waste of time as well. I'm complaining about the companies doing the rebranding, i.e. Twitter in this case.

hishamco commented:
Hopefully this is not a matter for Lombiq one day :)

davidpuplava commented:
> Zoltan I think we did similar things in both Azure AD & Facebook Meta

Yes, looks like UI strings were updated to use the Meta and EntraID names, but the type names were left as is, mostly. I'll work towards @Piedone instructions to put the type names back and focus on the UI Strings.

My interest in the X/Twitter module is because I implemented the v2 endpoint for posting to X. I'll submit that as a PR after I take care of this one.

The v2 API for X is substantial enough that I could clone the Twitter module and make a new `OrchardCore.X` module like @sebastienros suggested. But with pricing tiers of $100 / $5000 / $42000 per month, I don't know how much demand for twitter integration their is out there. Their API urls still use 'tweets', so they're not yet fully committed to their own rebrand.

davidpuplava commented:
This PR is ready for review. I changed type/class/id names back to their original names prior to the rebrand based on what I could find in #15528 but I left the changes to the documentation pages. I tested both workflows for both new and old Update Twitter Status tasks, and confirm they hit the v1.1 API for X/Twitter.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Twitter/Views/TwitterSettings.Edit.cshtml:
```diff
@@ -3,22 +3,22 @@
 <p class="alert alert-warning" role="alert">
     @T["The website might be restarted upon saving the settings, potentially leading to temporary unresponsiveness during the process."]
-    <a class="seedoc" href="@(OrchardCore.Admin.Constants.DocsUrl)reference/modules/Twitter/#twitter-integration" target="_blank">@T["See documentation"]</a>
+    <a class="seedoc" href="@(OrchardCore.Admin.Constants.DocsUrl)reference/modules/X/#-xtwitter-integration" target="_blank">@T["See documentation"]</a>
```
```suggestion
    <a class="seedoc" href="@(OrchardCore.Admin.Constants.DocsUrl)reference/modules/X/#x-twitter-integration" target="_blank">@T["See documentation"]</a>
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Twitter/Workflows/Startup.cs:
```diff
@@ -1,4 +1,5 @@
 using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Data.Migration;
```
```suggestion
```

Piedone commented:
Would you like to also review this, @hishamco?

hishamco commented:
I will do a quick one now
